### ISSUE TITLE
HTML / RST output generator

### ISSUE THREAD

**Issue by cputans** [HunterMcGushion/docstr_coverage#30]
Before submitting a PR with this functionality, want to confirm that adding the ability to generate ReStructuredText (for use with Sphinx) or HTML output of the coverage information is desired.

**Comment by MiWeiss** [comment]
Let me post my opinion on this:

**TL;DR: I am afraid that implementing HTML/RST/.... outputs would decrease the stability of `docstr-coverage`. A simple XML/JSON output might become simple to add in the future.** 

Detailed arguments:

1. **use case** I am having trouble thinking about central use-cases of this feature. In test-coverage, nicely visualized reports are critical to identifying test weaknesses, which is hard to do on text-based reports. `docstr-coverage` is much simpler and I do not think such a report would really be appreciated by many. Moreover, if the use-case is not 100% clear, it will be hard to come up with an appropriate design of these html/rst pages. @cputans maybe you can you explain for what you would use such a report?
2. **maintainability** `docstr-coverage` is a rather slim library. This is great for stability and maintainability. Adding full support for nice HTML / RST / ... outputs would require the addition of quite a large junk of new code. This makes maintenance harder and has the potential of decreasing stability, a price I would not want to pay for a feature with no obvious use-case.
3. **potential compromise** Instead of html/rst, I am having fewer troubles imagining a file-based output using a somewhat more structured format (e.g. xml or json). This would allow parsing the output of `docstr-coverage` by machines and thus allowing all kinds of fancy use-cases. At the same time, as there's no creativity required, such an implementation would be simpler and easier to maintain. In a different repository, someone could release a nice xml-to-html converter, fitting exactly their use-case.
4. **attack order** There's the following TODO currently in the code `Add option to sort reports by filename, coverage score... (ascending/descending)`. Such an option would require to collect all results during a run of `docstr-coverage` in one data structure. Once implemented, this would also allow easy implementation of an xml or json report as defined in *potential compromise*. From my personal use-cases, I consider the *sort-order* TODO quite important, especially for big projects with very high coverage (which currently requires console scrolling to identify missing docs) and there's a good chance it will be implemented next yet.  Implementing the xml/json output, if at all, should probably happen afterwards, once the central collection of `docstr-coverage` is implemented.

As mentioned above, this is just my opinion - and I'm not the code owner. So please feel free to argue with me :-)

**Comment by MiWeiss** [comment]
I'm proceeding with closing this issue for the reasons described above. 

Feel free to comment if you disagree - I'll happily reopen...

**Comment by Paebbels** [comment]
@MiWeiss I would like to deliver a use case.

We have lot's of Python projects (mainly library packages) which utilize:
* code coverage (coverage.py)
* static type checking (mymp)
* unit testing and performance testing(pytest)
* automatic documentation (Sphinx)
* ...

All is executed with GitHub Actions and published to GitHub Pages. Besides the Sphinx documentation, we also upload and publish the code coverage and static type checking reports to GH Pages. These are sub directories and we link them from the documentation navigation bar.

We use coverage percentages to check and also publish our code quality measures. We would like to achieve &ge;90 % for branch coverage.

Like we would like to keep code coverage high, we also need a good API documentation. Currently, I'm looking into docstr_coverage how it can be utilized for such a report that could be integrated too. But for now, there is no HTML or ReST output, so it can't be integrated. There is also no structured text file (e.g. YAML or JSON) for post-processing.

I think a HTML report could compress the code of a module just to declarations for coloring like this:

```
CONST = 42

def myFunction():
  # doc-string

class MyClass:
  # doc-string

  def myFunction():
    # doc-string
```

No need to render the full doc-string nor the code in such a report. Maybe the doc-string could repeat the first line, so the comment or doc-string doesn't look different from real code.

**Comment by Paebbels** [comment]
I did some prototyping and got this ...

![image](https://user-images.githubusercontent.com/956109/162641842-4e9d4a4d-27d3-4b03-9ae8-f71a795bf106.png)

while working on that code I found several limitations of `docstr_coverage`:
* Statistics are only collected per module.  
  A per class information is not available.
  * items per class
  * documented/undocumented per class
  * coverage per class
  * missing items per class
* `get_docstring_coverage` does not return a well-defined data model, because per file information is mixed with overall information.

**Comment by MiWeiss** [comment]
Hi @Paebbels and thanks for your comments!

Some comments:

1.  **Use Case** Do I understand your use-case correctly: The main purpose is that you want to show the documentation coverage to users (or to developers?) in the browser? Why would they want to know? (I'm not skeptical of your use case, I just want to make sure I understand its motivation in detail).
2. **Per-Class information** This is correct. If we want to change that, I think it should be on a report level (see below), but not for the overall functionality.
3. **`get_docstring_coverage` data model** Absolutely correct! We keep this data model for legacy reasons. Try `docstr_coverage.analyze`  for a newer data model (which is still experimental and especially the returned data structure will be extended as we add more features do docstr-coverage.

Overall context:
About a year ago, we refactored the code s.t. docstr-coverage now actually keeps the coverage state in memory (thus after this issue was closed) which now allows to implement the export of file reports much more easily (essentially, serializing the result of the analyze call). I have been planning to add a PR for that but did not get to do it just yet. I guess with your issue, I’ll get back to it soon. In the end, it should lead to something where the following call leads to a generated json report (and similar for other formats): `docstr-coverage <other options> --outputs report.json`

What regard the html output, I see two options:
1.	Add a serialization to html as well, i.e., allowing the call 
`docstr-coverage <other options> --outputs report.html` (actually, we'll probably need a folder)
2.	Implement a standalone library which converts aboves report.json into a nice html output. 

IMO, (1) is really only an option if we keep the report very simple (something along the line of your print screen). Otherwise, if we include too much style and other stuff, such a feature would increase technical depth too much. Something more complicated, like an extraction of the file and class hierarchy you proposed above would fall in the second category.

Do my thought make sense to you? In case we decide to implement option (1), would you be willing to contribute? If so, we can dive deeper into the design decisions.

**Comment by Paebbels** [comment]
@MiWeiss for an abstract and internal data model I was thinking more about like such:

(I just did some quick modelling)

```py
from typing import Dict


class SimpleMember:
	_name: str
	_isDocumented: bool
	
	@property
	def Name(self) -> str:
		return self._name
	
	@property
	def IsDocumented(self) -> bool:
		return self._isDocumented

	@property
	def Coverage(self) -> float:
		return 1.0 if self._isDocumented else 0.0


class CompositeMember(SimpleMember):
	_totalMembers: int
	_documentedMembers: int
	_undocumentedMembers: int
	
	@property
	def TotalMembers(self) -> int:
		return self._totalMembers
	
	@property
	def DocumentedMembers(self) -> int:
		return self._documentedMembers
	
	@property
	def UndocumentedMembers(self) -> int:
		return self._undocumentedMembers

	
class Variable(SimpleMember):
	pass
	

class Function(SimpleMember):
	pass


class Class(CompositeMember):
	_variables: Dict[str, Variable]
	_methods: Dict[str, Function]


class Package(CompositeMember):
	_variables: Dict[str, Variable]
	_methods: Dict[str, Function]
	_classes: Dict[str, Class]

```

**Comment by MiWeiss** [comment]
Excuse my late answer - have not noticed yours...

To get more fine-grained information, similar to what you proposed above, call `docstr_coverage.analyze` from your python code. That should give you most / all you need.

Note that the interface of `docstr_coverage.analyze` is experimental and will change in the future (likely, by providing even more info)
